modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php

```diff
@@ -77,36 +77,67 @@ public function enqueue_tracking_script() {
 	}
 
 	/**
-	 * On product lists or other non-product pages, add an event listener to "Add to Cart" button click
+	 * Render common default properties as a js args string.
 	 */
-	public function loop_session_events() {
+	public function get_common_properties() {
 		$blogid = Jetpack::get_option( 'id' );
 
-		// check for previous add-to-cart cart events
+		return "'blog_id': '" . esc_js( $blogid ) . "',
+			'ui': '" . esc_js( $this->get_user_id() ) . "',
+			'url': '" . esc_js( home_url() ) . "',
+			'woo_version': '" . esc_js( WC()->version ) . "',";
+	}
+
+	/**
+	 * Record an event with optional custom properties.
+	 *
+	 * @param string  $event_name The name of the event to record.
+	 * @param integer $product_id The id of the product relating to the event.
+	 * @param array   $properties Optional array of (key => value) event properties.
+	 */
+	public function record_event( $event_name, $product_id, $properties = array() ) {
+		$product = wc_get_product( $product_id );
+		if ( ! $product instanceof WC_Product ) {
+			return;
+		}
+		$product_details = $this->get_product_details( $product );
+
+		$extra_properties = '';
+		foreach ( $properties as $key => $value ) {
+			$extra_properties = $extra_properties . "'$key': '" . esc_js( $value ) . "', ";
+		}
+
+		wc_enqueue_js(
+			"_wca.push( {
+					'_en': '" . esc_js( $event_name ) . "',
+					'pi': '" . esc_js( $product_id ) . "',
+					'pn': '" . esc_js( $product_details['name'] ) . "',
+					'pc': '" . esc_js( $product_details['category'] ) . "',
+					'pp': '" . esc_js( $product_details['price'] ) . "',
+					'pt': '" . esc_js( $product_details['type'] ) . "'," .
+					$extra_properties . $this->get_common_properties() . '
+				} );'
```

Comment:
We're doing a lot of string concatenation and repeated calls to `esc_json()`. That preserves the previous coding style, but I think it would be more readable and maintainable to pass arrays around, with the values unescaped.

When the string finally has to be generated (in `event_push()`, if we go with that approach, or else here,) that would be the moment to escape everything with an `array_walk()` call, and then encode with `json_encode()`. That would mean we need to review where single vs. double quotes are being used, wherever we output this.

If we were working with arrays before output, then `get_common_properties()` could be passing this a raw array, without doing part of the escaping there, and part in the various other places it gets called.

Reply:
I agree it would be nice to have a single path for all these events. The issue is that those two other events are generated on the front end, and pull data from the DOM, so they call the JS api directly - we can't enqueue a pre-rendered `_wca.push()` call.

In future, we could render a `event_push()` style wrapper func on the front end which supplies all the default props. I didn't want to go that far with the refactor :)